Nettoyage de printemps du Gulpfile, partie 2

Situphen opened the pull request:
- Améliore les tâches liées au sprite
  - Les deux fichiers `sprite.png` et `sprite@2x.png` sont minimisées dans une tâche séparée du reste pour réduire le temps de build
  - Les images servant à créer le sprite ne sont plus minimisées car elles ne nous servent à rien unitairement
- Suppression de JSHint et mise en place de ESLint avec une configuration basée sur StandardJS
  - Dépendances en `~x.y.z` dans le `package.json` car les versions mineures peuvent modifier des règles mais pas les patchs
  - `yarn run lint` pour vérifier si tout est bon
  - `yarn run lint --fix` pour corriger ce qui peut l'être automatiquement
- On garde en mémoire la version de Node JS dans le `.nvmrc` et non pas
dans `scripts/define_variable.sh`
  - Si jamais on décide de changer la version de Node, on la change dans le `.nvmrc`
  - Si la version installée diffère de celle dans le `.nvmrc`, `nvm` râle ce qui permet d'avertir le développeur qu'il faut mettre à jour la version de Node avec `nvm install`
- Adaptation des fichiers JS à la configuration de ESLint
  - Quelques corrections manuelles nécessaires
  - Corrections automatiques avec `--fix` pour le reste

**QA :**

- Activer l'environnement Python avec `source zdsenv/bin/activate`
- Normalement `nvm` devrait râler et c'est normal
- Mettre à jour la version de Node avec `nvm install`
- Mettre à jour les dépendances Node avec `yarn`
- Vérifier les commandes `yarn run build` et `yarn run watch` pour ce qui concerne le sprite
- Vérifier les commandes `yarn run lint` et `yarn run lint --fix`
- Vérifier que le site fonctionne bien avec `make run-back`

Situphen committed:
Améliore (encore) les tâches liées au sprite

- Minification des fichiers sprite.png et sprite@2x.png
- Rapidité du build

Situphen committed:
Changement de linter : adieu JSHint, bienvenue ESLint

A-312 commented on assets/js/modal.js:
```diff
@@ -17,206 +17,205 @@
+        }).join(' ')
+      }
+
+      Modal({
```
Tu as fais une très grosse modification ici, es-tu certain ?

`this` ne va pas retourner pareil sans

Situphen replied:
Je t'avoue que j'ai eu l'erreur `error  Do not use 'new' for side effects         no-new` donc j'ai supprimé le `new` en espérant que ça ne change rien au fonctionnement :D

A-312 replied:
Rajoute une variable devant, ça devrait faire l'affaire

A-312 commented on Gulpfile.js:
```diff
@@ -1,146 +1,160 @@
+const terser = require('gulp-terser-js')
+
+
+/// / Speed mode
```
C'est bizarre cette façon de faire pourquoi pas : 
```js
// Speed mode
```

ou

```js
/*
 * Speed mode
 */
```

A-312 commented on assets/js/content-publication-readiness.js:
```diff
@@ -1,20 +1,23 @@
+    var readiness = $(e.target).data('is-ready').toString() === 'true'
+    var csrf = $('input[name=csrfmiddlewaretoken]').val()
+    var toggledReadiness = !readiness
+    $.ajax(url, { method: 'PUT',
```
``method: 'PUT',`` à la ligne ?

A-312 commented on assets/js/content-publication-readiness.js:
```diff
@@ -1,20 +1,23 @@
+      },
+      headers: {
+        'X-CSRFToken': csrf
+      } })
```
à la ligne

Situphen commented:
J'ai presque fini mon nettoyage, il y a juste deux petits points qui me dérangent avec la configuration actuelle :

- les commentaires JSDOC sont modifiés (à voir si ça les rend invalides, en tous cas c'est moche) 
  ```js
  /**
   * Toggle the text of the element by the text of an attribute
   *
   * @param {string} dataAttribute
   */
  ```
  
  ```js
  /**
     * Toggle the text of the element by the text of an attribute
     *
     * @param {string} dataAttribute
     */
  ```
- les longues lignes n'émettent pas d'erreur ou d'avertissement (ça on peut le modifier) et ne peuvent pas être modifiées (peu m'importe le style tant que c'est plus court) automatiquement par le linter (avec la commande `yarn run lint --fix`)
  ```js
  gulp.watch(['assets/images/sprite/*.png', 'assets/scss/_sprite.scss.hbs'], gulp.series(cssSprite, gulp.parallel(css, imagesSprite)))
  ```
  
  ```js
  gulp.watch(
    ['assets/images/sprite/*.png', 'assets/scss/_sprite.scss.hbs'],
    gulp.series(cssSprite, gulp.parallel(css, imagesSprite))
  )
  ```

Je vais aussi essayer de reprendre la documentation du *frontend*, mais ce sera dans une autre PR pour éviter que celle-ci traine.

A-312 commented:
```
gulp.watch(
  ['assets/images/sprite/*.png', 'assets/scss/_sprite.scss.hbs'],
  gulp.series(cssSprite, gulp.parallel(css, imagesSprite))
)
```

C'est correct ça ? Utilise deux variables

Situphen commented:
J'ai mis à jour ma branche, j'ai réparé les JSDocs et je laisse tomber l'histoire des longues lignes (ça se fera au cas par cas lors de la QA ce n'est pas grave). Il y a peut-être encore quelques ajustements à faire puis on pourra passer à la QA.

Situphen committed:
Adaptation manuelle des fichiers au nouveau linter

Situphen committed:
On garde en mémoire la version de Node JS dans le .nvmrc

On garde en mémoire la version de Node JS dans le .nvmrc et non pas
dans scripts/define_variable.sh. Ainsi, lorsqu'on modifie la version
de Node JS, les développeurs sont avertis du changement par nvm et
mettent à jour Node JS.

Situphen commented:
C'est enfin prêt à être QAtisé !

AmauryCarrade commented:
Alors. **QA OK !** J'ai juste remarqué quelques détails précisés ci-dessous mais je les pense bénins.

> Mettre à jour les dépendances Node avec `yarn`

`yarn` a râlé sur des co-dépendances non-satisfaites, mais ça n'a rien perturbé après et il ne s'agissait que d'avertissements.

<details><summary>Logs de <code>yarn</code></summary>
<pre>$ yarn               
yarn install v1.21.1
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.2.11: The platform "linux" is incompatible with this module.
info "fsevents@1.2.11" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > eslint-config-standard@14.1.0" has unmet peer dependency "eslint@>=6.2.2".
warning " > eslint-plugin-import@2.19.1" has unmet peer dependency "eslint@2.x - 6.x".
warning " > eslint-plugin-jsdoc@19.1.0" has unmet peer dependency "eslint@^5.0.0 || ^6.0.0".
warning " > eslint-plugin-node@11.0.0" has unmet peer dependency "eslint@>=5.16.0".
warning "eslint-plugin-node > eslint-plugin-es@3.0.0" has unmet peer dependency "eslint@>=4.19.1".
warning " > eslint-plugin-standard@4.0.1" has unmet peer dependency "eslint@>=5.0.0".
[5/5] Building fresh packages...
Done in 18.93s.</pre></details>

> Vérifier les commandes `yarn run build` et `yarn run watch` pour ce qui concerne le sprite

Tout est OK, les sprites sont correctement créées et mises à jour (testé en renommant des images dans le dossier `assets/images/sprite`). Aucun problème, le `_sprite.scss` est bien mis à jour. Idem en ajoutant une nouvelle image. Autant en `watch` qu'en `build`.

(Bien sûr renommer une image fait planter l'étape du SCSS vu que ça renomme une variable avec, mais osef)

> Vérifier les commandes `yarn run lint` et `yarn run lint --fix`

Y'avait rien à corriger mais ça n'a pas retourné d'erreur.

<details><summary>Logs de <code>yarn run lint [--fix]</code></summary>
<pre>$ yarn run lint
yarn run v1.21.1
$ node ./node_modules/gulp/bin/gulp.js lint
The speed mode is not enabled.
[17:12:03] Using gulpfile ~/Sites/zds-site/Gulpfile.js
[17:12:03] Starting 'lint'...
[17:12:04] Finished 'lint' after 1.31 s
Done in 2.20s.</pre>

<pre>$ yarn run lint --fix
yarn run v1.21.1
$ node ./node_modules/gulp/bin/gulp.js lint --fix
The speed mode is not enabled.
[17:12:13] Using gulpfile ~/Sites/zds-site/Gulpfile.js
[17:12:13] Starting 'lint'...
[17:12:14] Finished 'lint' after 1.32 s
Done in 2.27s.</pre></details>

> Vérifier que le site fonctionne bien avec `make run-back`

Aucun soucis. De même avec `make run` (par acquit de conscience).

Situphen commented:
> `yarn` a râlé sur des co-dépendances non-satisfaites, mais ça n'a rien perturbé après et il ne s'agissait que d'avertissements.

J'ai eu la même chose. J'ai l'impression qu'il suit l'ordre des paquets tels qu'ils sont dans `package.json` et installe les `eslint-config-*` et `eslint-plugin-*` avant `eslint` donc lors de l'installation de ces paquets il émet un avertissement.

AmauryCarrade commented:
> J'ai eu la même chose. J'ai l'impression qu'il suit l'ordre des paquets tels qu'ils sont dans `package.json` et installe les `eslint-config-*` et `eslint-plugin-*` avant `eslint` donc lors de l'installation de ces paquets il émet un avertissement.

Ça explique pourquoi ça ne pose aucun soucis après, du coup. Parfait. :)

Situphen commented:
Je viens de mettre à jour la branche par rapport à `dev` mais sans aucune modification.

-------

@artragis Vu que la PR modifie tout le front, j'imagine que l'idéal c'est de la merger juste après avoir fini la v29. Mentionne moi quand c'est le cas comme ça je pourrais mettre à jour la branche, puis envoyé c'est pesé.

A-312 commented:
Une raison pour passer à NPM si ça fonctionne mieux 🤔

Situphen committed:
Ajoute un point-virgule entre les fichiers JS

Situphen committed:
Met à jour les dépendances NodeJS

Situphen commented:
J'ai rebasé, mis à jour le `yarn.lock` et j'ai aussi corrigé le bout de code du Gulpfile qui met un `;` entre les fichiers JS lorsqu'on les concatène.

philippemilink commented:
Rapport de QA
============

OK.

Mêmes remarques que Amaury. Je n'ai pas réussi à faire cracher une erreur à `yarn run lint` en alignant mal une ligne dans un fichier JS. Je ne sais si c'est normal, ou si le linter se met à râler seulement pour d'autres choses.

Situphen committed:
Mise à jour du yarn.lock

Situphen committed:
Adaptation automatique des fichiers au nouveau linter

Situphen commented:
J'ai mis à jour ma branche par rapport à `upstream/dev`.

> Je n'ai pas réussi à faire cracher une erreur à yarn run lint en alignant mal une ligne dans un fichier JS. Je ne sais si c'est normal, ou si le linter se met à râler seulement pour d'autres choses.

Ce n'est pas normal ! Il me met une erreur quand j'indente mal une ligne (que ce soit un multiple de 4 ou pas). À noter que pour le fichier `assets/js/editor.js` est exclu du lint, cette PR n'a que trop tardé donc je le ferai dans une autre PR.

```
/home/situphen/Code/zds-site/assets/js/autocompletion.js
  307:1  error  Expected indentation of 4 spaces but found 6  indent

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

philippemilink commented:
Je testais justement avec le fichier `editor.js` ! Je viens de tester avec le fichier `accordeon.js` et c'est bon, `yarn lint` râle bien, `--fix` corrige bien les erreurs.

**QA ok**

Situphen commented:
Travis est vert, merge qui veut !
